test/unit/org/apache/cassandra/service/replication/migration/KeyspaceMigrationInfoTest.java

```diff
@@ -0,0 +1,382 @@
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.dht.IPartitioner;
+import org.apache.cassandra.dht.Murmur3Partitioner;
+import org.apache.cassandra.dht.NormalizedRanges;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.io.util.DataInputBuffer;
+import org.apache.cassandra.io.util.DataOutputBuffer;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.tcm.Epoch;
+import org.apache.cassandra.tcm.membership.NodeVersion;
+
+import static org.junit.Assert.*;
+
+public class KeyspaceMigrationInfoTest
+{
+    private static IPartitioner partitioner;
+    private static TableId testTableId;
+
+    @BeforeClass
+    public static void setup() throws Exception
+    {
+        System.setProperty(CassandraRelevantProperties.PARTITIONER.getKey(), Murmur3Partitioner.class.getName());
+
+        ServerTestUtils.prepareServerNoRegister();
+        partitioner = DatabaseDescriptor.getPartitioner();
+        assertTrue(partitioner instanceof Murmur3Partitioner);
+        testTableId = TableId.generate();
+    }
+
+    @Test
+    public void testConstruction()
+    {
+        Epoch epoch = Epoch.create(1);
+        Map<TableId, NormalizedRanges<Token>> pendingRangesPerTable = Collections.emptyMap();
+
+        KeyspaceMigrationInfo info = new KeyspaceMigrationInfo(
+            "test_ks",
+            pendingRangesPerTable,
+            epoch
+        );
+
+        assertEquals("test_ks", info.keyspace);
+        assertEquals(epoch, info.startedAtEpoch);
+        assertTrue(info.isComplete());
+        assertTrue(info.pendingRangesPerTable.isEmpty());
+    }
+
+    @Test
+    public void testWithRangesRepairedForTable()
+    {
+        List<Range<Token>> ranges = createTestRanges();
+        Epoch epoch1 = Epoch.create(1);
+        Epoch epoch2 = Epoch.create(2);
+
+        // Start with full ring as pending for testTableId
+        Range<Token> fullRing = new Range<>(partitioner.getMinimumToken(), partitioner.getMinimumToken());
+        NormalizedRanges<Token> fullRingNormalized = NormalizedRanges.normalizedRanges(Collections.singleton(fullRing));
+        Map<TableId, NormalizedRanges<Token>> pendingRangesPerTable = Collections.singletonMap(testTableId, fullRingNormalized);
+
+        KeyspaceMigrationInfo info = new KeyspaceMigrationInfo(
+            "test_ks",
+            pendingRangesPerTable,
+            epoch1
+        );
+
+        // Subtract first range - should leave remaining ranges as pending
+        KeyspaceMigrationInfo updated = info.withRangesRepairedForTable(testTableId, Collections.singleton(ranges.get(0)));
+
+        assertFalse(updated.isComplete());
+
+        // Subtract second range - should leave even fewer pending ranges
+        KeyspaceMigrationInfo updated2 = updated.withRangesRepairedForTable(testTableId, Collections.singleton(ranges.get(1)));
+
+        assertFalse(updated2.isComplete());
```

Comment:
This isn't asserting much about what the outcome should be? Nothing could have changed and this test would pass.